File: app/graph/signals.py
"""Signal detection for market moves and volume spikes."""

import os
from datetime import datetime, timezone

from app.graph.state import SignalTier, SignalType, TickerSignal

PRICE_MINOR = float(os.getenv("SIGNAL_PRICE_MINOR", "3.0"))
PRICE_MAJOR = float(os.getenv("SIGNAL_PRICE_MAJOR", "5.0"))
VOLUME_MINOR = float(os.getenv("SIGNAL_VOLUME_MINOR", "1.5"))
VOLUME_MAJOR = float(os.getenv("SIGNAL_VOLUME_MAJOR", "3.0"))


def classify_signal(
    day_change_pct: float,
    volume_ratio: float,
    price_minor: float = PRICE_MINOR,
    price_major: float = PRICE_MAJOR,
    volume_minor: float = VOLUME_MINOR,
    volume_major: float = VOLUME_MAJOR,
) -> dict | None:
    """Classify a ticker's current data into a signal, or None if below thresholds."""
    abs_change = abs(day_change_pct) if day_change_pct else 0
    vol_r = volume_ratio if volume_ratio else 0

    price_hit = abs_change >= price_minor
    volume_hit = vol_r > volume_minor
    price_major_hit = abs_change >= price_major
    volume_major_hit = vol_r >= volume_major

    if price_hit and volume_hit:
        return {
            "signal_type": SignalType.COMBINED,
            "tier": SignalTier.MAJOR,
            "value": abs_change,
        }

    if price_major_hit:
        return {
            "signal_type": SignalType.PRICE_MOVE,
            "tier": SignalTier.MAJOR,
            "value": abs_change,
        }

    if volume_major_hit:
        return {
            "signal_type": SignalType.VOLUME_SPIKE,
            "tier": SignalTier.MAJOR,
            "value": vol_r,
        }

    if price_hit:
        return {
            "signal_type": SignalType.PRICE_MOVE,
            "tier": SignalTier.MINOR,
            "value": abs_change,
        }

    if volume_hit:
        return {
            "signal_type": SignalType.VOLUME_SPIKE,
            "tier": SignalTier.MINOR,
            "value": vol_r,
        }

    return None


def detect_signals_for_ticker(
    snapshot: dict,
    prev_snapshot: dict | None,
) -> list[TickerSignal]:
    """Detect signals for a single ticker given current and previous snapshot."""
    ticker = snapshot.get("ticker", "")
    day_change_pct = snapshot.get("day_change_pct")
    volume = snapshot.get("volume")

    if day_change_pct is None and volume is None:
        return []

    volume_ratio = 0.0
    if volume and prev_snapshot and prev_snapshot.get("volume"):
        volume_ratio = volume / prev_snapshot["volume"]

    result = classify_signal(
        day_change_pct=day_change_pct or 0,
        volume_ratio=volume_ratio,
    )
    if result is None:
        return []

    return [
        TickerSignal(
            ticker=ticker,
            signal_type=result["signal_type"],
            tier=result["tier"],
            value=result["value"],
            detected_at=datetime.now(timezone.utc),
        )
    ]

File: app/src/graph/signals.ts
// app/src/graph/signals.ts
import 'dotenv/config'
import type { TickerSignal, SignalType, SignalTier } from './state'

function getThresholds() {
  return {
    priceMinor: Number(process.env.SIGNAL_PRICE_MINOR ?? 3.0),
    priceMajor: Number(process.env.SIGNAL_PRICE_MAJOR ?? 5.0),
    volMinor:   Number(process.env.SIGNAL_VOLUME_MINOR ?? 1.5),
    volMajor:   Number(process.env.SIGNAL_VOLUME_MAJOR ?? 3.0),
  }
}

interface SignalResult {
  signal_type: SignalType
  tier: SignalTier
  value: number
}

export function classifySignal(dayChangePct: number, volumeRatio: number): SignalResult | null {
  const t = getThresholds()
  const absPct = Math.abs(dayChangePct)

  const priceTriggered = absPct >= t.priceMinor
  const volTriggered = volumeRatio >= t.volMinor
  const priceMajor = absPct >= t.priceMajor
  const volMajor = volumeRatio >= t.volMajor

  if (!priceTriggered && !volTriggered) return null

  let signal_type: SignalType
  if (priceTriggered && volTriggered) signal_type = 'COMBINED'
  else if (priceTriggered) signal_type = 'PRICE_MOVE'
  else signal_type = 'VOLUME_SPIKE'

  const tier: SignalTier = (priceMajor || volMajor || signal_type === 'COMBINED') ? 'MAJOR' : 'MINOR'
  const value = priceTriggered ? absPct : volumeRatio

  return { signal_type, tier, value }
}

/** Drop signals whose ticker was signal-analyzed within SIGNAL_COOLDOWN_MIN.
 *  Without this, a stock pinned at a big day move re-triggers MAJOR analysis
 *  every runner cycle and starves the scheduled cadence. */
export function filterCooledSignals(
  signals: TickerSignal[],
  cooldowns: Record<string, string>,
  nowMs: number = Date.now(),
): TickerSignal[] {
  const cooldownMin = Number(process.env.SIGNAL_COOLDOWN_MIN ?? 60)
  return signals.filter(s => {
    const last = cooldowns[s.ticker]
    if (!last) return true
    return (nowMs - new Date(last).getTime()) / 60_000 >= cooldownMin
  })
}

export function detectSignalsForTicker(
  ticker: string,
  snap: { day_change_pct: number | null; volume: number | null },
  prevSnap?: { volume: number | null } | null,
): TickerSignal[] {
  const dayChangePct = snap.day_change_pct ?? 0
  const volumeRatio = snap.volume && prevSnap?.volume && prevSnap.volume > 0
    ? snap.volume / prevSnap.volume
    : 0

  const result = classifySignal(dayChangePct, volumeRatio)
  if (!result) return []

  return [{
    ticker,
    signal_type: result.signal_type,
    tier: result.tier,
    value: result.value,
    detected_at: new Date().toISOString(),
  }]
}
